windows/README_MINGW

```diff
@@ -0,0 +1,216 @@
+mv mingw64 /
+
+Make sure GCC still runs and it is the version we need:
+
+$ gcc --version
+gcc.exe (x86_64-posix-sjlj-rev0, Built by MinGW-W64 project) {version}
+
+
+Install JACK depdendencies
+--------------------------
+
+There are prebuilt MinGW binaries for all the libraries that can be installed
+using the pacman package manager, but since there are many MinGW flavors it is
+better idea to build from source to avoid any linker and runtime issues.
+Fortunately there are PKGBUILD files for doing so together with a nice guide at
+https://github.com/msys2/MINGW-packages
+
+git clone https://github.com/msys2/MINGW-packages.git
+
+cd MINGW-packages/mingw-w64-{libname}-git
+MINGW_INSTALLS=mingw64 makepkg-mingw -sLf
+pacman -U mingw-w64-{libname}-git-{suffix}.pkg.tar.xz
+
+Repeat procedure for each library replacing {libname} with the appropriate name
+and {suffix} with whatever the above process created. For example:
+mingw-w64-x86_64-libtre-git-r128.6fb7206-2-any.pkg.tar.xz
+
+db
+libtre
+portaudio
+libsndfile
+libsamplerate
+
+Some libraries like libsndfile and libsamplerate will ask for installing extra
+dependencies, it is ok to do so.
+
+
+Compile JACK
+------------
+
+- Clone repo
+git clone https://github.com/jackaudio/jack2
+cd jack2
+
+- (Needs better solution) Allow gcc to find regex.h and db.h, pkgconfig is not
+of much help here:
+
+export CPATH=$CPATH:/mingw64/include/tre:/mingw64/include
+
+- (Needs wscript solution) disable check for db.h otherwise waf will configure
+without metadata support. Edit wscript and comment out the following line:
+
+db.check(header_name='db.h')
+
+Note that currently metadata support is not working on Windows and probably
+involves patching common/JackMetadata.cpp:
+server\__db.001: No such file or directory
+
+Creating the mentioned "server" directory results in:
+Cannot open metadata DB at server\metadata.db: No such file or directory
+
+So for now it is recommended to pass --db=no to waf configure or leaving wscript
+intact, so the db.check() call fails and DB ends disabled. Leaving DB enabled
+will make jackd behave erratically or crash.
+
+- (Needs wscript solution) define missing constants, this can be done by editing
+build/config.h:
+
+#define ADDON_DIR "jack"
+#define PRIu64 "llu"     (unsigned long long is 64bits wide on MinGW-W64)
+//#define HAVE_ASIO 1    (do not use this yet) 
+
+Do not forget to re-add these lines after every waf configure run. As stated
+above, a final solution should come from wscript.
+
+- Link libjack to libstdc++ statically. Some client apps like Ardour already
+ship with a libstdc++-6.dll library that will take precedence, resulting in
+missing symbols during runtime load of libjack.dll. Add the following link flags
+to common/wscript @ ~200:
+
+if bld.env['IS_WINDOWS']:
+    clientlib.env.append_value('LINKFLAGS', ['-static-libstdc++', '--disable-auto-import'])
+
+- Finally build and install
+
+./waf configure --prefix=/opt/jack --db=no
+./waf
+./waf install
+
```

Comment:
Unfortunately can not get past this, any suggestions please, have setup asio as you have defined

```
$ ./waf configure --prefix=/opt/jack --db=no
Setting top to                           : /home/BSFau/jack2
Setting out to                           : /home/BSFau/jack2/build
Checking for 'clang++' (C++ compiler)    : not found
Checking for 'g++' (C++ compiler)        : /mingw64/bin/g++
Checking for 'clang' (C compiler)        : not found
Checking for 'gcc' (C compiler)          : /mingw64/bin/gcc
Checking platform                        : Windows
Checking for ASIO SDK                    : yes
Checking for program 'pkg-config'        : /mingw64/bin/pkg-config
Checking for 'alsa >= 1.0.18'            : not found
Checking for 'libffado >= 1.999.17'      : not found
Checking for 'gtkIOStream >= 1.4.0'      : not found
Checking for 'eigen3 >= 3.1.2'           : not found
Checking for header windows.h            : yes
Checking for 'portaudio-2.0 >= 19'       : yes
Checking for header mmsystem.h           : yes
Checking for 'celt >= 0.11.0'            : not found
Checking for 'celt >= 0.8.0'             : not found
Checking for 'celt >= 0.7.0'             : not found
Checking for 'celt >= 0.5.0'             : not found
Checking for header opus/opus_custom.h   : not found
Checking for 'opus >= 0.9.0'             : not found
Checking for 'samplerate'                : yes
Checking for 'sndfile'                   : yes
Checking for library readline            : not found
Checking for header readline/readline.h  : not found
Checking for header systemd/sd-daemon.h  : not found
Checking for library systemd             : not found
Checking for header alloca.h             : not found
Checking for ppoll                       : no
Checking for header execinfo.h           : not found
Invalid argument 'function_name' in test
Checking for header regex.h              : yes
Invalid argument 'function_name' in test
Checking for header winsock2.h           : yes
Invalid argument 'function_name' in test
Checking for header ['windows.h', 'mmsystem.h'] : yes
Invalid argument 'function_name' in test
Checking for header ['windows.h', 'psapi.h']    : yes
Checking for ucontext->uc_mcontext.gp_regs      : no
Checking for ucontext->uc_mcontext.uc_regs      : no
Checking for ucontext->uc_mcontext.mc_gregs     : no
Checking for ucontext->uc_mcontext.gregs        : no
Checking for NGREG                              : no

==================
JACK 1.9.13 svn revision will checked and eventually updated during build
Maximum JACK clients                            : 64
Maximum ports per application                   : 768
Install prefix                                  : /opt/jack
Library directory                               : /opt/jack/lib
Drivers directory                               : /opt/jack/bin/jack
Build debuggable binaries                       : no
C compiler flags                                : ['-Wall']
C++ compiler flags                              : ['-Wall', '-std=gnu++11']
Linker flags                                    : ['-Wl,--enable-auto-import', '-Wl,--enable-auto-import']
Build with engine profiling                     : no
Build with 32/64 bits mixed mode                : no
Build standard JACK (jackd)                     : yes
Build D-Bus JACK (jackdbus)                     : no
Autostart method                                : classic
Build doxygen documentation                     : no
Enable ALSA driver                              : no
Enable FireWire driver (FFADO)                  : no
Enable IIO driver                               : no
Enable Portaudio driver                         : yes
Enable WinMME driver                            : yes
Build with CELT                                 : no
Build Opus netjack2                             : no
Build with libsamplerate                        : yes
Build with libsndfile                           : yes
Build with readline                             : no
Use systemd notify                              : no
Use Berkeley DB (metadata)                      : no

'configure' finished successfully (14.038s)

BSFau@BSF-PC-AU1 MSYS ~/jack2
$ ./waf -v
Waf: Entering directory `/home/BSFau/jack2/build'
[138/223] Linking build/common/audioadapter.dll
00:24:09 runner ['/mingw64/bin/g++', '-Wl,--enable-auto-import', '-Wl,--enable-auto-import', '-Wl,--out-implib,common/audioadapter.dll.a', '-shared', 'common/JackResampler.cpp.7.o', 'common/JackLibSampleRateResampler.cpp.7.o', 'common/JackAudioAdapter.cpp.7.o', 'common/JackAudioAdapterInterface.cpp.7.o', 'common/JackAudioAdapterFactory.cpp.7.o', 'windows/portaudio/JackPortAudioAdapter.cpp.7.o', 'windows/portaudio/JackPortAudioDevices.cpp.7.o', '-o/home/BSFau/jack2/build/common/audioadapter.dll', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-Lcommon', '-LD:/msys64/mingw64/lib', '-LD:/msys64/mingw64/lib', '-ljackserver', '-lpthread', '-lregex', '-lws2_32', '-lpsapi', '-lwinmm', '-lsamplerate', '-lportaudio', '-lwinmm', '-lm', '-ldsound', '-lole32', '-lwinmm', '-lm', '-luuid', '-lsetupapi', '-lole32', '-lwinmm', '-lm', '-lole32', '-luuid', '-lwinmm', '-lm', '-lole32', '-luuid']
[155/223] Linking build/jack_portaudio.dll
00:24:09 runner ['/mingw64/bin/g++', '-Wl,--enable-auto-import', '-Wl,--enable-auto-import', '-Wl,--out-implib,portaudio.dll.a', '-shared', '-shared', 'windows/portaudio/JackPortAudioDevices.cpp.8.o', 'windows/portaudio/JackPortAudioDriver.cpp.8.o', '-o/home/BSFau/jack2/build/jack_portaudio.dll', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-Lcommon', '-LD:/msys64/mingw64/lib', '-ljackserver', '-lportaudio', '-lwinmm', '-lm', '-ldsound', '-lole32', '-lwinmm', '-lm', '-luuid', '-lsetupapi', '-lole32', '-lwinmm', '-lm', '-lole32', '-luuid', '-lwinmm', '-lm', '-lole32', '-luuid', '-lpthread', '-lregex', '-lws2_32', '-lpsapi', '-lwinmm']
windows/portaudio/JackPortAudioDevices.cpp.8.o:JackPortAudioDevices.cpp:(.text+0x1821): undefined reference to `PaAsio_GetAvailableBufferSizes'
windows/portaudio/JackPortAudioDevices.cpp.8.o:JackPortAudioDevices.cpp:(.text+0x1bfd): undefined reference to `PaAsio_GetAvailableBufferSizes'
windows/portaudio/JackPortAudioDriver.cpp.8.o:JackPortAudioDriver.cpp:(.text+0x11e3): undefined reference to `PaAsio_GetInputChannelName'
windows/portaudio/JackPortAudioDriver.cpp.8.o:JackPortAudioDriver.cpp:(.text+0x1316): undefined reference to `PaAsio_GetOutputChannelName'
collect2.exe: error: ld returned 1 exit status

windows/portaudio/JackPortAudioDevices.cpp.7.o:JackPortAudioDevices.cpp:(.text+0x1821): undefined reference to `PaAsio_GetAvailableBufferSizes'
windows/portaudio/JackPortAudioDevices.cpp.7.o:JackPortAudioDevices.cpp:(.text+0x1bfd): undefined reference to `PaAsio_GetAvailableBufferSizes'
collect2.exe: error: ld returned 1 exit status

Waf: Leaving directory `/home/BSFau/jack2/build'
Build failed
 -> task in 'portaudio' failed with exit status 1:
        {task 7696569732816: cxxshlib JackPortAudioDevices.cpp.8.o,JackPortAudioDriver.cpp.8.o -> jack_portaudio.dll,portaudio.dll.a}
['/mingw64/bin/g++', '-Wl,--enable-auto-import', '-Wl,--enable-auto-import', '-Wl,--out-implib,portaudio.dll.a', '-shared', '-shared', 'windows/portaudio/JackPortAudioDevices.cpp.8.o', 'windows/portaudio/JackPortAudioDriver.cpp.8.o', '-o/home/BSFau/jack2/build/jack_portaudio.dll', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-Lcommon', '-LD:/msys64/mingw64/lib', '-ljackserver', '-lportaudio', '-lwinmm', '-lm', '-ldsound', '-lole32', '-lwinmm', '-lm', '-luuid', '-lsetupapi', '-lole32', '-lwinmm', '-lm', '-lole32', '-luuid', '-lwinmm', '-lm', '-lole32', '-luuid', '-lpthread', '-lregex', '-lws2_32', '-lpsapi', '-lwinmm']
 -> task in 'audioadapter' failed with exit status 1:
        {task 7696569673680: cxxshlib JackResampler.cpp.7.o,JackLibSampleRateResampler.cpp.7.o,JackAudioAdapter.cpp.7.o,JackAudioAdapterInterface.cpp.7.o,JackAudioAdapterFactory.cpp.7.o,JackPortAudioAdapter.cpp.7.o,JackPortAudioDevices.cpp.7.o -> audioadapter.dll,audioadapter.dll.a}
['/mingw64/bin/g++', '-Wl,--enable-auto-import', '-Wl,--enable-auto-import', '-Wl,--out-implib,common/audioadapter.dll.a', '-shared', 'common/JackResampler.cpp.7.o', 'common/JackLibSampleRateResampler.cpp.7.o', 'common/JackAudioAdapter.cpp.7.o', 'common/JackAudioAdapterInterface.cpp.7.o', 'common/JackAudioAdapterFactory.cpp.7.o', 'windows/portaudio/JackPortAudioAdapter.cpp.7.o', 'windows/portaudio/JackPortAudioDevices.cpp.7.o', '-o/home/BSFau/jack2/build/common/audioadapter.dll', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-Lcommon', '-LD:/msys64/mingw64/lib', '-LD:/msys64/mingw64/lib', '-ljackserver', '-lpthread', '-lregex', '-lws2_32', '-lpsapi', '-lwinmm', '-lsamplerate', '-lportaudio', '-lwinmm', '-lm', '-ldsound', '-lole32', '-lwinmm', '-lm', '-luuid', '-lsetupapi', '-lole32', '-lwinmm', '-lm', '-lole32', '-luuid', '-lwinmm', '-lm', '-lole32', '-luuid']

```

Reply:
Is it possible your portaudio is missing ASIO support? the binary version that can be installed via pacman lacks ASIO. The guide contains instructions to build portaudio from the MinGW source package, its PKGBUILD file has to be patched to enable ASIO.

Reply:
I believe I have patched  this successfully
![image](https://user-images.githubusercontent.com/11743864/69498664-3be64080-0f4f-11ea-81bf-d601dab38aff.png)

I checked the build logs and to my relatively inexperienced eye it looked ok, but on checking again now i see in configuration summary `ASIO = no`' I will look further but if I have missed something please let me know. Thanks for your huge effort with this

![image](https://user-images.githubusercontent.com/11743864/69498730-f413e900-0f4f-11ea-9130-f4f4128af3e2.png)

Reply:
FIXED, FOUND it, my error I only added ASIO support to the build static,

can we use git apply with a patch instead of individually hand editing.  maybe something that can be scripted, but forgive me if this is not valid I am very much a  NOOB with mingW and C++,C let alone cross platform.

@lucianoiam  this PR is an awesome attempt  at simplifying the build, well done

by the way for anyone else hitting this a git diff should look like this
```
diff --git a/mingw-w64-portaudio/PKGBUILD b/mingw-w64-portaudio/PKGBUILD
index 4daa221b7..7ee63ee0c 100644
--- a/mingw-w64-portaudio/PKGBUILD
+++ b/mingw-w64-portaudio/PKGBUILD
@@ -53,10 +53,12 @@ build() {
     --disable-shared \
     --enable-static \
     --with-dxdir=${MINGW_PREFIX}/${MINGW_CHOST} \
-    --with-winapi=wmme,directx,wasapi,wdmks
+         --with-asiodir=/opt/asiosdk  \
+    --with-winapi=wmme,directx,wasapi,wdmks,asio

   make
-
+  find /opt/asiosdk -name "*.lo" -type f -delete
+  find /opt/asiosdk -name "*.o" -type f -delete
   [[ -d "build-shared-${MINGW_CHOST}" ]] && rm -rf "build-shared-${MINGW_CHOST}"
   mkdir -p "${srcdir}/build-shared-${MINGW_CHOST}"
   cd "${srcdir}/build-shared-${MINGW_CHOST}"
@@ -68,7 +70,8 @@ build() {
     --target=${MINGW_CHOST} \
     --enable-shared \
     --with-dxdir=${MINGW_PREFIX}/${MINGW_CHOST} \
-    --with-winapi=wmme,directx,wasapi,wdmks
+         --with-asiodir=/opt/asiosdk  \
+    --with-winapi=wmme,directx,wasapi,wdmks,asio

   make
 }

```

Reply:
That is great! this patch makes clear what has to be done in PKGBUILD. I think a clean future proof solution is submitting a patch to the MinGW package mantainers that turns on ASIO support by default for portaudio if the required dependencies are present, without hardcoding a path or properly telling the user where to place ASIO.

Reply:
I'm not entirely sure about this but I think it's safer to run
```bash
find /opt/asiosdk -name "*.lo" -type f -delete
find /opt/asiosdk -name "*.o" -type f -delete
```
also after the last `make`, as the temp files from the last build could cause issues when re-running this with the first build. What do you think?

Reply:
You are right, I am updating the guide now. Stay tuned for the next commit, it also contains a patch to allow compiling 1.9.14 on MinGW. Thanks.